feat(tag): introducing custom icon

IgnacioBecerra opened the pull request:
Related: https://github.com/carbon-design-system/carbon-for-ibm-dotcom/issues/3669

This is a contribution for the `Tag` component in which allows the adopter to include any icon from `@carbon/icons` before the content text. This feature had been requested as part of the Expressive theme in Carbon for IBM.com. 

The new `renderIcon` prop requires an icon component as an argument, which will render it accordingly within the `Tag` component. `hasCustomIcon` must also be passed into the `Tag` component to activate the custom icon functionality.

Here are some examples:

Using `renderIcon: Tag16`
<img width="123" alt="Screen Shot 2020-11-11 at 10 00 00 AM" src="https://user-images.githubusercontent.com/24970122/98847279-e24b8000-2404-11eb-9220-dc06d302d0aa.png">


Using `renderIcon: Compass16`
<img width="121" alt="Screen Shot 2020-11-11 at 10 00 40 AM" src="https://user-images.githubusercontent.com/24970122/98847287-e4adda00-2404-11eb-8e82-560837e7c76b.png">





#### Changelog

**New**

- added `renderIcon` prop to allow for a custom icon to be added before the content text


#### Testing / Reviewing

Enable/disable the new prop, experiment with various icon names to ensure they render properly.

IgnacioBecerra committed:
feat(tag): introducing custom icon

IgnacioBecerra committed:
feat(tag): add test

IgnacioBecerra committed:
feat(tag): fixed variable name

jeanservaas commented:
Hey @IgnacioBecerra thanks for the contribution!

I was just looking at Shixie's spec I think I would tweak the spacing a bit 

She's actually modified our tag for digital (.com team) and is using: 

- 14/20pt type
- 32px tag height (6px of padding top and bottom)
- Gray 80 (your type is too light for contrast accessibility right now)
**- I changed her spec to make the space between the icon and the type 4px (instead of 8px)**

<img width="260" alt="image" src="https://user-images.githubusercontent.com/43144260/98737010-3e61c600-236b-11eb-89cb-13f4d193d7ad.png">

If you wanted a spec to add an icon to the Carbon tag, for product teams (which it looks like you're doing), here it is:

- [x] 12/16pt type
- [x] 24px tag height (6px of padding top and bottom)
- [ ] Gray 80 (your type is too light for contrast accessibility right now)
- [ ] Make the space between the icon and the type 4px (instead of 8px)

<img width="286" alt="image" src="https://user-images.githubusercontent.com/43144260/98737799-5be35f80-236c-11eb-9f01-8922f60824ec.png">

IgnacioBecerra committed:
Merge branch 'master' into tag-icon-feature

tw15egan commented on packages/react/src/components/Tag/Tag.js:
```diff
@@ -9,7 +9,8 @@ import PropTypes from 'prop-types';
 import classNames from 'classnames';
 import { settings } from 'carbon-components';
-import { Close16 } from '@carbon/icons-react';
+import * as Icons from '@carbon/icons-react';
```
I think it may be best to render icons using similar to the `renderIcon` prop found in other components

See Button and Tooltip for examples: 
https://github.com/carbon-design-system/carbon/blob/0049b31dfdfb54774eb49ab17f0df1b900cfa963/packages/react/src/components/Button/Button.js

https://github.com/carbon-design-system/carbon/blob/225f212a8f3e2bd86d4a23f374f205b25bf5b9f0/packages/react/src/components/Tooltip/Tooltip.js

mjabbink commented:
Icon cropping in the left corner


![96F8F21B-C620-4A13-85DD-EC5A8017C8AB](https://user-images.githubusercontent.com/32881239/98740699-1e350580-2371-11eb-9e4c-b824a3bb8a0c.png)

IgnacioBecerra committed:
feat(tag): improving design based on feedback

IgnacioBecerra commented:
@jeanservaas Thank you for the specs! I modified the text color values to type 80 in `tokens.scss` to conform with contrast accessibility guidelines. 

@tw15egan Thank you for the examples! Using the `renderIcon` prop, an icon component will be passed as a prop into the `Tag` component instead of using strings. The Storybook knob for custom icons will be gone however, although it's not strictly necessary as the other components lack it as well.

@mjabbink Thanks for letting me know about that! I wasn't able to replicate the cropping, but I removed the css property that I suspect was the cause of the issue.

IgnacioBecerra committed:
feat(tag): removed unecessary icon import

tw15egan commented:
@IgnacioBecerra Awesome! If you do want to allow a user to see that they can change the icon to whatever they want, you can add in a knob, like how we do for the button story. Here's [the link](https://github.com/carbon-design-system/carbon/blob/master/packages/react/src/components/Button/Button-story.js#L18-L57) if you're interested

IgnacioBecerra committed:
feat(tag): added custom icon knob to story

tw15egan commented on packages/react/src/components/Tag/Tag.js:
```diff
@@ -9,7 +9,7 @@ import PropTypes from 'prop-types';
 import classNames from 'classnames';
 import { settings } from 'carbon-components';
-import { Close16 } from '@carbon/icons-react';
+import { Close16 }from '@carbon/icons-react';
```
```suggestion
import { Close16 } from '@carbon/icons-react';
```

IgnacioBecerra committed:
feat(tag): fixed spacing on imports

IgnacioBecerra commented:
@tw15egan Thanks! Added three examples in the Storybook using the `select` knob.

tw15egan commented on packages/react/src/components/Tag/Tag.js:
```diff
@@ -34,6 +34,8 @@ const Tag = ({
   type,
   filter,
+  hasCustomIcon,
+  renderIcon: CustomIconElement,
```
Would we expect a user to always set their own custom icon? If so, do we just need the `renderIcon` prop, and base the rest of the logic on the presence of this prop?

IgnacioBecerra replied:
Yes, users are expected to set their own custom icon if they choose to use this `Tag` variation, there's no "default" icon. Users would also need to include the `hasCustomIcon` in their `Tag` component, but yeah, I think needing two variables to activate this behavior may be redundant. 

Would you suggest changing it to only needing `renderIcon`/getting rid of `hasCustomIcon`?

tw15egan replied:
Yeah, that's what I was thinking! That way by default there would be no icon and if a user wants to use one they would be expected to provide their own icon. Then we can just check if `renderIcon` exists

IgnacioBecerra replied:
Got it, pushed the change, thanks!

IgnacioBecerra committed:
feat(tag): removed redundant boolean

tw15egan commented on packages/react/src/components/Tag/Tag.js:
```diff
@@ -117,6 +139,12 @@ Tag.propTypes = {
   onClose: PropTypes.func,
 
+    /**
+   * Optional prop to allow overriding the icon rendering.
```
```suggestion
   * Optional prop to render a custom icon.
```

tw15egan commented on packages/react/src/components/Tag/Tag-test.js:
```diff
@@ -64,6 +64,11 @@ describe('Tag', () => {
   });
 
+  it('should allow for a custom icon', () => {
+    const tag = shallow(<Tag type="red" customIcon='Add16'>This is a tag</Tag>)
```
```suggestion
    const tag = shallow(<Tag type="red" renderIcon={Add16}>This is a tag</Tag>)
```

tw15egan replied:
Will also need to import the icon at the top of the test file

```
import { Add16 } from '@carbon/icons-react';
```

IgnacioBecerra committed:
Update packages/react/src/components/Tag/Tag-test.js

Co-authored-by: TJ Egan <tw15egan@gmail.com>

IgnacioBecerra committed:
Update packages/react/src/components/Tag/Tag.js

Co-authored-by: TJ Egan <tw15egan@gmail.com>

tw15egan committed:
Merge branch 'master' into tag-icon-feature

tw15egan committed:
chore(snapshot): update snapshots, run prettier

tw15egan reviewed (approved):
LGTM, thanks for taking the time to work on this! 👍 ✅ 🚀

emyarod commented on packages/components/src/components/tag/_tag.scss:
```diff
@@ -172,6 +172,22 @@
+    flex-shrink: 0;
+    width: rem(16px);
+    height: rem(16px);
+    margin-right: rem(4px);
```
```suggestion
    margin-right: $spacing-02;
```

emyarod commented on packages/react/src/components/Tag/Tag-story.js:
```diff
@@ -36,6 +49,15 @@ const props = {
+      onClick: action('onClick'),
+      onClose: action('onClose'),
+      renderIcon: iconToUse,
+    };
```
I think we can just inline this since we aren't checking for falsiness or the svgData property

```suggestion
    return {
      ...this.regular(),
      onClick: action('onClick'),
      onClose: action('onClose'),
      renderIcon: iconMap[select('Icon (icon)', icons, 'Tag16')],
    };
```

emyarod commented on packages/react/src/components/Tag/Tag.js:
```diff
@@ -76,6 +78,25 @@ const Tag = ({
         <Close16 />
       </button>
     </div>
+  ) : CustomIconElement ? (
```
can we refactor this to avoid nesting ternaries?

IgnacioBecerra committed:
feat(tag): refactored nested ternary

IgnacioBecerra committed:
Merge branch 'tag-icon-feature' of https://github.com/IgnacioBecerra/carbon into tag-icon-feature

IgnacioBecerra commented:
@emyarod Thanks. for the suggestions! I refactored using if/else if for better readability, let me know if you have any thoughts about it!

emyarod commented on packages/react/src/components/Tag/Tag.js:
```diff
@@ -50,40 +51,67 @@ const Tag = ({
+        </button>
+      </div>
+    );
+  } else if(CustomIconElement) {
```
```suggestion
  }
  if (CustomIconElement) {
```

since we `return` in the block above, this following `else-if` block can just stand alone as an `if` block, if that makes sense

emyarod commented on packages/react/src/components/Tag/Tag.js:
```diff
@@ -50,40 +51,67 @@ const Tag = ({
+      </div>
+    );
+  } else {
+    return (
```
```suggestion
  }
  return (
```

emyarod commented on packages/react/src/components/Tag/Tag.js:
```diff
@@ -50,40 +51,67 @@ const Tag = ({
-    </span>
-  );
+    );
+  }
```
```suggestion
```

removed because the final `return` statement does not need to be in an `else` block

emyarod reviewed (commented):
it looks like you are not using an autoformatter in your editor, can you try running `yarn format` in the repository root?

emyarod commented on packages/react/src/components/Tag/Tag.js:
```diff
@@ -50,40 +51,67 @@ const Tag = ({
+        <button
+          className={`${prefix}--tag__custom-icon`}>
+          <CustomIconElement />
+        </button>
```
```suggestion
        <button
          type="button"
          className={`${prefix}--tag__custom-icon`}>
          <CustomIconElement />
        </button>
```

I believe this also needs a `type` attribute

IgnacioBecerra replied:
Hmm, thinking about it, the custom icon isn't a button in this context (it can't be clicked). Closing the `CustomIconElement` in a `<div>` would provide the same effect, so I'll be changing it to that instead.

jeanservaas reviewed (approved):
lgtm

IgnacioBecerra committed:
feat(tag): reformatted conditionals

IgnacioBecerra committed:
Merge branch 'master' into tag-icon-feature

tw15egan committed:
Merge branch 'master' into tag-icon-feature

joshblack commented on packages/react/src/components/Tag/Tag.js:
```diff
@@ -51,32 +52,60 @@ const Tag = ({
+        }
+        id={tagId}
+        {...other}>
+        <button className={`${prefix}--tag__custom-icon`}>
```
Would this button need the same props as the other one?

On a related note, would it make sense to try and re-use the markup and conditionally render the custom icon?

tw15egan replied:
Does this even need to be a button since this is just a container for the icon and isn't clickable?

IgnacioBecerra replied:
Seems like I didn't push the change yesterday, but the custom icon shouldn't behave like a button. I had changed it to a `div` instead, but I'm currently looking into the conditional rendering instead of having three blocks of code in there

joshblack commented on packages/react/src/components/Tag/Tag.js:
```diff
@@ -117,6 +146,12 @@ Tag.propTypes = {
+   * Optional prop to render a custom icon.
+   * Can be a React component class
+   */
+  renderIcon: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
```
Would there be a reasonable way to just allow teams to render the icon by passing it as a child to the component? Wasn't sure but wanted to check 👀

IgnacioBecerra replied:
It's possible to do what you described, although it would lack the styles required from the specs, spacing wise.

IgnacioBecerra committed:
feat(tag): refactored icon rendering code

IgnacioBecerra committed:
feat(tag): ran formatter

IgnacioBecerra committed:
Merge branch 'master' into tag-icon-feature

joshblack commented on packages/react/src/components/Tag/Tag.js:
```diff
@@ -77,12 +79,18 @@ const Tag = ({
+      ) : (
+        ''
+      )}
+      <span title={typeof children === 'string' ? children : null} {...other}>
```
Noticed that `{...other}` was being spread to both the outer `<div>` and the `<span>` here, I think we'll need to pick one and stick with it otherwise we'll end up in weird situations like both are receiving an `onClick` or `data-testid` or something 🤔 

I think just keeping `{...other}` on the outermost node would work, but curious what you think!

IgnacioBecerra replied:
I agree, keeping it on the outermost node works better. Good catch, thanks @joshblack!

IgnacioBecerra committed:
feat(tag): removed inner span's ...other options

joshblack committed:
Merge branch 'master' into tag-icon-feature

joshblack reviewed (approved):
Thanks for doing this!

kodiakhq[bot] committed:
Merge branch 'master' into tag-icon-feature
